File: crates/lucidos-app/src/api/client/releaseNotices.ts
import { API, json, mutatingFetch, throwIfNotOk } from './_core';

/** One *release notice*: something this release needs the reader to know or do.
 *
 *  Authored in the repo and baked into the engine, so the text is the same on
 *  every workspace running this version. Distinct from a *release note*, which
 *  says what changed; a notice says what you have to do about it. */
export interface ReleaseNotice {
  /** Stable forever. The workspace remembers what it answered by this id. */
  id: string;
  /** The release this applies FROM, plain semver. A floor, not a stamp: the
   *  engine hides the notice until it reports this version or newer. */
  since: string;
  title: string;
  /** Raw markdown. Rendered client-side (`utils/renderMarkdown.ts`). */
  body: string;
  /** The button's label. Present exactly when `action_prompt` is. */
  action_label?: string;
  /** The sentence the button SENDS as a new message. */
  action_prompt?: string;
  /** True once this workspace has answered it. The panel keeps showing it. */
  resolved: boolean;
}

/** Everything both surfaces are drawn from, in one response so they cannot
 *  disagree about what is answered. */
export interface ReleaseNoticeView {
  /** Every notice this release has reached, oldest first. */
  notices: ReleaseNotice[];
  /** The one the modal shows, or `null` when the workspace owes nothing. */
  next_id: string | null;
}

/** See engine `GET /api/v1/release-notices`. */
export async function releaseNotices(): Promise<ReleaseNoticeView> {
  return json(`${API}/release-notices`);
}

/** Record that the user answered `id`, and get the settled list back.
 *
 *  Answering one the workspace already walked past changes nothing rather than
 *  failing: two devices showing the same modal is ordinary. See engine
 *  `POST /api/v1/release-notices/resolve`. */
export async function resolveReleaseNotice(id: string): Promise<ReleaseNoticeView> {
  const resp = await mutatingFetch(`${API}/release-notices/resolve`, {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify({ id }),
  });
  await throwIfNotOk(resp);
  return resp.json();
}

File: crates/lucidos-engine/src/api/release_notices.rs
//! HTTP for *release notices*: what this workspace still owes the reader, and
//! the answer that settles one.
//!
//! Two endpoints, one list. The modal draws the notice named by `next_id`, and
//! the What's New panel draws them all. Both read this one response, so they
//! cannot disagree about what is answered.
//!
//! - `GET  /api/v1/release-notices`
//! - `POST /api/v1/release-notices/resolve` with body `{ "id": "<notice id>" }`
//!
//! The rules live in `engine::release_notices`. This module is the boundary.

use axum::extract::State;
use axum::{
    http::HeaderMap,
    routing::{get, post},
    Json, Router,
};
use serde::Deserialize;

use super::error::ApiError;
use super::AppState;
use crate::engine::event_bus::SystemEvent;
use crate::engine::release_notices;

/// The release this engine reports, which decides what a caller may see or
/// answer.
///
/// Read here rather than taken from the request. A client could otherwise name
/// a release it is not on, and be handed a notice that has not shipped.
fn running_release() -> Result<semver::Version, ApiError> {
    semver::Version::parse(crate::LUCIDOS_RELEASE)
        .map_err(|e| ApiError::internal(format!("{} is not semver: {e}", crate::LUCIDOS_RELEASE)))
}

/// `GET /api/v1/release-notices`.
async fn list(
    State(state): State<AppState>,
) -> Result<Json<release_notices::NoticeView>, ApiError> {
    let running = running_release()?;
    let cursor = release_notices::stored_cursor(&state.pool).await;
    Ok(Json(release_notices::view(
        release_notices::all(),
        &running,
        cursor.as_deref(),
    )))
}

#[derive(Deserialize)]
struct ResolveRequest {
    id: String,
}

/// `POST /api/v1/release-notices/resolve`.
///
/// Answering a notice the workspace has already walked past is a no-op rather
/// than an error. Two devices showing the same modal is ordinary, so the second
/// answer changes nothing and announces nothing.
async fn resolve(
    State(state): State<AppState>,
    headers: HeaderMap,
    Json(body): Json<ResolveRequest>,
) -> Result<Json<release_notices::NoticeView>, ApiError> {
    let running = running_release()?;
    let notices = release_notices::all();
    if !notices.iter().any(|n| n.id == body.id) {
        return Err(ApiError::not_found(format!(
            "no release notice with id {:?} in this build",
            body.id
        )));
    }
    let moved = release_notices::resolve(&state.pool, notices, &running, &body.id)
        .await
        .map_err(ApiError::db)?;
    if moved {
        state
            .engine
            .event_bus
            .emit_user_system(
                &headers,
                &state.pool,
                "[ReleaseNotices] ReleaseNoticeResolved",
                |actor| SystemEvent::ReleaseNoticeResolved {
                    notice_id: body.id.clone(),
                    actor,
                },
            )
            .await;
    }
    list(State(state)).await
}

pub(super) fn router() -> Router<AppState> {
    Router::new()
        .route("/release-notices", get(list))
        .route("/release-notices/resolve", post(resolve))
}
